pymc/distributions/multivariate.py

```diff
@@ -1513,78 +1517,83 @@ def make_node(self, rng, size, n, eta):
+        )(outer_rng, scan_rng, size, n, eta)
 
     @classmethod
     def _random_corr_matrix(
-        cls, rng: Variable, n: int, eta: TensorVariable, flat_size: TensorVariable
+        cls,
+        outer_rng: Variable,
+        scan_rng: Variable,
+        n: int,
+        eta: TensorVariable,
+        size: TensorVariable,
     ) -> tuple[Variable, TensorVariable]:
-        # original implementation in R see:
-        # https://github.com/rmcelreath/rethinking/blob/master/R/distributions.r
+        size_is_none = rv_size_is_none(size)
+        size = () if size_is_none else size
 
-        beta = eta - 1.0 + n / 2.0
-        next_rng, beta_rvs = pt.random.beta(
-            alpha=beta, beta=beta, size=flat_size, rng=rng
+        beta0 = eta - 1.0 + n / 2.0
+
+        outer_rng_out, y0 = pt.random.beta(
+            alpha=beta0, beta=beta0, size=size, rng=outer_rng
         ).owner.outputs
-        r12 = 2.0 * beta_rvs - 1.0
-        P = pt.full((flat_size, n, n), pt.eye(n))
-        P = P[..., 0, 1].set(r12)
-        P = P[..., 1, 1].set(pt.sqrt(1.0 - r12**2))
-        n = get_underlying_scalar_constant_value(n)
-        for mp1 in range(2, n):
-            beta -= 0.5
-            next_rng, y = pt.random.beta(
-                alpha=mp1 / 2.0, beta=beta, size=flat_size, rng=next_rng
+
+        r12 = 2.0 * y0 - 1.0
+
+        P0 = pt.full((*size, n, n), pt.eye(n))
+        P0 = P0[..., 0, 1].set(r12)
+        P0 = P0[..., 1, 1].set(pt.sqrt(1.0 - r12**2))
+
+        def step(mp1, beta, P, prev_rng):
+            beta_next = beta - 0.5
+
+            middle_rng, y = pt.random.beta(
+                alpha=mp1 / 2.0, beta=beta, size=size, rng=prev_rng
             ).owner.outputs
+
             next_rng, z = pt.random.normal(
-                loc=0, scale=1, size=(flat_size, mp1), rng=next_rng
+                loc=0, scale=1, size=(*size, mp1), rng=middle_rng
             ).owner.outputs
-            z = z / pt.sqrt(pt.einsum("ij,ij->i", z, z.copy()))[..., np.newaxis]
-            P = P[..., 0:mp1, mp1].set(pt.sqrt(y[..., np.newaxis]) * z)
+
+            ein_sig_z = "i, i->" if z.ndim == 1 else "...ij, ...ij->...i"
+
+            z = z / pt.sqrt(pt.einsum(ein_sig_z, z, z.copy()))[..., None]
+            P = P[..., 0:mp1, mp1].set(pt.sqrt(y[..., None]) * z)
             P = P[..., mp1, mp1].set(pt.sqrt(1.0 - y))
-        C = pt.einsum("...ji,...jk->...ik", P, P.copy())
-        return next_rng, C
 
+            return (beta_next, P), {prev_rng: next_rng}
 
-class MultivariateIntervalTransform(Interval):
-    name = "interval"
+        (_, P_seq), updates = pytensor.scan(
+            fn=step,
+            outputs_info=[beta0, P0],
+            sequences=[pt.arange(2, n)],
+            non_sequences=[scan_rng],
+            strict=True,
+        )
 
-    def log_jac_det(self, *args):
-        return super().log_jac_det(*args).sum(-1)
+        P = pt.concatenate([P0[None], P_seq], axis=0)[-1]
```

Comment:
Whay do you concatenate and then take the last value? Sometimes P_seq is empty?

Reply:
I already have it, it's `P0`. The issue isn't how to get the initial state, it's how to give it back when (at runtime) n=2.